examples/color.ts

```diff
@@ -0,0 +1,133 @@
+import * as bnb from "../src/bread-n-butter";
+import { prettyPrint } from "./util";
+
+interface HasLimit {
+  upperLimit: number;
+  lowerLimit: number;
+}
+
+interface HasBoundaryCheck {
+  boundaryCheck: () => asserts this is this;
+}
+
+class Transparency implements HasLimit, HasBoundaryCheck {
+  constructor(public value: number) {}
+  get upperLimit() {
+    return 1.0;
+  }
+  get lowerLimit() {
+    return 0.0;
+  }
+  boundaryCheck() {
+    if (this.upperLimit < this.value || this.lowerLimit > this.value) {
+      throw new Error("invalid value of transparency");
+    }
+  }
+}
+
+class RGBValue implements HasLimit, HasBoundaryCheck {
+  constructor(public value: number) {}
+  get upperLimit() {
+    return 0;
+  }
+  get lowerLimit() {
+    return 255;
+  }
+  boundaryCheck() {
+    if (
+      this.upperLimit < this.value ||
+      this.lowerLimit > this.value ||
+      !Number.isInteger(this.value)
+    ) {
+      throw new Error("invalid RGB value");
+    }
+  }
+}
+
+class Color implements HasBoundaryCheck {
+  constructor(
+    public r: RGBValue,
+    public g: RGBValue,
+    public b: RGBValue,
+    public a: Transparency | null
+  ) {}
```

Comment:
I would like to focus more on the parser in this example, so maybe less class code?

Just the class Color like:

```ts
class Color {
  constructor(public r: number, public g: number, public b: number, public a: number = 1)  {}
  toString() { ... }
}
```